stub-pthreads/barrier.c

```diff
@@ -0,0 +1,23 @@
+#include "pthread_impl.h"
+/*
+	Note on PTHREAD_PROCESS_SHARED:
+	Because WASM doesn't have virtual memory nor subprocesses,
+	there isn't any way for the same synchronization object to have multiple mappings.
+	Thus we can completely ignore it here.
+*/
+
+int pthread_barrier_init(pthread_barrier_t *restrict b, const pthread_barrierattr_t *restrict a, unsigned count)
+{
+	if (count-1 > INT_MAX-1) return EINVAL;
+	*b = (pthread_barrier_t){ ._b_limit = count-1 };
+	return 0;
+}
+int pthread_barrier_destroy(pthread_barrier_t *b)
+{
+	return 0;
+}
+int pthread_barrier_wait(pthread_barrier_t *b)
+{
+	if (!b->_b_limit) return PTHREAD_BARRIER_SERIAL_THREAD;
+	__builtin_trap();
+}
```

Comment:
Let's look at it again:

- `pthread_barrier_init` creates a barrier in a way appropriate for a single-threaded environment.
- `pthread_barrier_destroy` destroys a barrier in a way appropriate for a single-threaded environment, which in this case happens to do nothing.
-`pthread_barrier_wait` waits on a barrier in a way appropriate for a single-threaded environment, i.e. traps since this would result in a deadlock, except for the 1-thread case where it returns `PTHREAD_BARRIER_SERIAL_THREAD`.

If I didn't miss anything, this is a conformant single-threaded implementation, so there is nothing that "doesn't work" there.

Reply:
I see what you mean; let me look at this again.